File: bidding-engine/src/main/java/com/omnibus/bidding/rules/DroolsCompilationService.java
package com.omnibus.bidding.rules;

import com.omnibus.bidding.domain.ManagedRuleDefinition;
import java.util.List;
import org.kie.api.KieServices;
import org.kie.api.builder.KieBuilder;
import org.kie.api.builder.KieFileSystem;
import org.kie.api.builder.Message;
import org.kie.api.runtime.KieContainer;
import org.springframework.stereotype.Service;

@Service
public class DroolsCompilationService {

    public KieContainer buildContainer(List<ManagedRuleDefinition> ruleDefinitions) {
        KieServices kieServices = KieServices.Factory.get();
        KieFileSystem kieFileSystem = kieServices.newKieFileSystem();

        for (ManagedRuleDefinition ruleDefinition : ruleDefinitions) {
            kieFileSystem.write(ruleDefinition.sourcePath(), ruleDefinition.content());
        }

        KieBuilder kieBuilder = kieServices.newKieBuilder(kieFileSystem).buildAll();
        boolean hasErrors = kieBuilder.getResults().hasMessages(Message.Level.ERROR);
        if (hasErrors) {
            throw new IllegalArgumentException(kieBuilder.getResults().toString());
        }

        return kieServices.newKieContainer(kieServices.getRepository().getDefaultReleaseId());
    }
}

File: bidding-engine/src/main/kotlin/com/omnibus/bidding/rules/DroolsCompilationService.kt
package com.omnibus.bidding.rules

import com.omnibus.bidding.domain.ManagedRuleDefinition
import org.kie.api.KieServices
import org.kie.api.builder.Message
import org.kie.api.runtime.KieContainer
import org.springframework.stereotype.Service

@Service
class DroolsCompilationService {
    fun buildContainer(ruleDefinitions: List<ManagedRuleDefinition>): KieContainer {
        val kieServices = KieServices.Factory.get()
        val kieFileSystem = kieServices.newKieFileSystem()

        for (ruleDefinition in ruleDefinitions) {
            kieFileSystem.write(ruleDefinition.sourcePath, ruleDefinition.content)
        }

        val kieBuilder = kieServices.newKieBuilder(kieFileSystem).buildAll()
        val hasErrors = kieBuilder.results.hasMessages(Message.Level.ERROR)
        if (hasErrors) {
            throw IllegalArgumentException(kieBuilder.results.toString())
        }

        return kieServices.newKieContainer(kieServices.repository.defaultReleaseId)
    }
}

